pkg/reconcilermanager/controllers/rootsync_controller_manager_test.go

```diff
@@ -78,46 +84,89 @@ func TestRootSyncReconcilerDeploymentLifecycle(t *testing.T) {
-		// keep watching
-		return errors.Errorf("reconciler deployment %s", event.Type)
+	// Start the watches before creating the RootSync.
+	// While this order is not required, it means we should see distinct Added vs Modified events when debugging.
+	tg := taskgroup.New()
+
+	t.Log("Starting deployment controller simulation")
+	reconcilerKey := core.RootReconcilerObjectKey(rs.Name)
+	tg.Go(func() error {
+		return simulateDeploymentController(ctx, t, fakeClient, reconcilerWatcher, reconcilerKey)
+	})
+
+	t.Log("Starting RootSync validator")
+	rsKey := client.ObjectKeyFromObject(rs)
+	tg.Go(func() error {
+		return validateRootSyncSetup(ctx, t, fakeClient, rsyncWatcher, rsKey)
 	})
+
+	t.Log("Creating RootSync")
+	err = fakeClient.Create(ctx, rs)
 	require.NoError(t, err)
-	if reconcilerObj == nil {
-		t.Fatal("timed out waiting for reconciler deployment to be applied")
+
+	t.Log("Waiting for Deployment Controller simulation & RootSync validation to stop")
+	if err := tg.Wait(); err != nil {
+		t.Fatal(err)
 	}
 
 	t.Log("verifying the reconciler-manager finalizer is present")
-	rsKey := client.ObjectKeyFromObject(rs)
 	rs = &v1beta1.RootSync{}
 	err = fakeClient.Get(ctx, rsKey, rs)
 	require.NoError(t, err)
 	require.True(t, controllerutil.ContainsFinalizer(rs, metadata.ReconcilerManagerFinalizer))
 
-	t.Log("deleting sync object and watching for NotFound")
-	err = watchutil.DeleteAndWait(ctx, fakeClient, rs, 10*time.Second)
+	// Simulate managed object deletion being blocked.
+	// TODO: Simulate deployment graceful shutdown, which is a much more likely use scenario.
+	// Unfortunately, to be able to simulate graceful shutdown, we would need
+	// to apply a patch to the deployment, but the current implementation of
+	// server-side apply in the fake client acts like an update, because the
+	// field-manager code is in an `internal` package upstream.
+	// However, the controller-manager has a forked version we can use.
+	// TODO: Update fakeClient.Patch to handle SSA with multiple field managers.
+	testFinalizer := "test-delete-blocked"
+	crbKey := reconcilerKey
+	saObj := &corev1.ServiceAccount{}
+	err = fakeClient.Get(ctx, crbKey, saObj)
+	require.NoError(t, err)
+	require.True(t, controllerutil.AddFinalizer(saObj, testFinalizer))
+	err = fakeClient.Update(ctx, saObj)
+	require.NoError(t, err)
+
+	t.Log("Deleting sync object")
+	err = fakeClient.Delete(ctx, rs)
+	require.NoError(t, err)
+
+	t.Log("Simulating teardown blocked")
+	time.Sleep(5 * time.Second)
+
+	// Validate ServiceAccount still exists (means fake client handles finalizers correctly)
+	err = fakeClient.Get(ctx, crbKey, saObj)
+	require.NoError(t, err)
+
+	// Validate RootSync still exists (means reconciler-manager finalizer blocks waiting for ServiceAccount not found)
+	err = fakeClient.Get(ctx, rsKey, rs)
+	require.NoError(t, err)
+
+	t.Log("Simulating teardown unblocked")
+	require.True(t, controllerutil.RemoveFinalizer(saObj, testFinalizer))
+	err = fakeClient.Update(ctx, saObj)
```

Comment:
Why updating the service account again after finalizer removal?

Reply:
RemoveFinalizer only modifies the local object. Gotta send the update to the server.

Reply:
Ack.